### ISSUE TITLE
Kill Checker Framework violation reports for findFirstToken

### ISSUE THREAD

**Issue by Gaoyan1999** [checkstyle/checkstyle#18804]
Follow-up: #18482 

In #18482, a `nullUntil` function is provided as a POC, which is a good start. However, after going through the codebase, I found that `findFirstToken` is used in many places. If we add `NullUtil.notNull(...)` everywhere to assert non-null, it will introduce a lot of redundant code that is hard to read, hard to maintain, and generally ugly.

My suggestion is to introduce utility wrappers around `findFirstToken`, so that we can eliminate Checker Framework violations in a centralized and cleaner way, instead of asserting non-null at every call site.

Below are several proposals.

---
### Background and Example
I will use findFirstToken(TokenTypes.IDENT) as an example:

After scanning the codebase, there are 100+ calls to `findFirstToken(TokenTypes.IDENT)`, which result in 87 Checker Framework errors. However, in most cases, this method cannot return null.

Currently, the only known case where findFirstToken(TokenTypes.IDENT) can return null is when dealing with a receiver parameter, as shown below

```java
public static boolean isReceiverParameter(DetailAST parameterDefAst) {
    return parameterDefAst.findFirstToken(TokenTypes.IDENT) == null;
}
```
From this, we can conclude that:

When calling `ast.findFirstToken(TokenTypes.IDENT)`, only parameterDefAst may return `null`

For other AST types, the result is effectively non-null

Therefore, my core idea is to stop calling `findFirstToken` directly, and instead route all calls through a wrapper that encodes this knowledge.

---
### Solution 1: Simple wrapper utility methods

Provide two utility methods in TokenUtil (or a similar utility class):

1. TokenUtil.findIdentChild(DetailAST ast) → @Nullable DetailAST

2. TokenUtil.requireIdentChild(DetailAST ast) → @NonNull DetailAST
    Throws an exception if the result is null

Cons:
This solution is simple, but it relies on developers knowing which method to use in which context. Misuse is still possible.

### Soluction 2: Typed AST “views” (personally preferred)

Since only parameterDefAst may return null, we can encode this behavior by introducing lightweight view classes for different AST types.

```java
// Current code
DetailAST ident = methodDef.findFirstToken(TokenTypes.IDENT); // @Nullable (from Checker's perspective)

// Solution 2
MethodDefView view = MethodDefView.of(methodDef);
DetailAST ident = view.getIdent(); // @NonNull, method definitions always have IDENT

ParameterDefView view = ParameterDefView.of(parameterDef);
DetailAST ident = view.getIdent(); // @Nullable, receiver parameter is possible

```
Advantages:
Once the AST type is known (which is true in most cases), the nullability of getIdent() becomes obvious

Checker Framework violations are resolved naturally via stronger type information

Call sites become cleaner and more expressive.

I also considered creating subclasses of DetailAST instead of using composition, since inheritance could provide even stronger type guarantees. However, that approach would require large-scale refactoring and changes across the codebase. Using composition via view objects feels like a reasonable trade-off: lighter, safer, and easier to adopt incrementally.

**Comment by Gaoyan1999** [comment]
Hi, @romani 
If you think this proposal is reasonable, I’m happy to work on it.

**Comment by romani** [comment]
@vivek-0509, please share your thoughts on this, while I busy with a lot of other stuff.

**Comment by vivek-0509** [comment]
As per my opinion
Regarding Solution 1 (simple wrapper utility methods like findIdentChild and requireIdentChild), I believe this approach does not add significant value over the existing NullUtil.notNull() pattern. It is essentially the same concept but specialized for one token type. If we were to adopt this approach, we would need to create similar wrapper methods for many other frequently used token types such as MODIFIERS, TYPE, PARAMETERS, and others. This would require creating alot of separate wrapper methods (like requireIdentChild, requireModifiersChild etc.), but all these methods would do the exact same thing that NullUtil.notNull() already does with just one generic method. 

Regarding Solution 2 (typed AST views using composition), while this is an elegant solution from a type-safety perspective, I believe it is over-engineered for the problem at hand. Implementing this approach would require creating view classes for numerous AST types including METHOD_DEF, CLASS_DEF, INTERFACE_DEF, RECORD_DEF, ENUM_DEF, and many more. This represents a significant implementation effort and would require refactoring a lot of call sites across the codebase. The adoption cost is very high, and developers would need to learn which view class to use in each context. If we implement the Views approach, it would create an ongoing maintenance burden, we do need to keep updating the view classes whenever Java adds new features or when we discover new edge cases in grammar rules. Given that the findFirstToken related suppressions represent only about low proportion of all Checker Framework suppressions, the return on investment for such a major architectural change does not seem justified which is already achieveable through simple approach of using NullUtil.notNull().

It iss also worth noting that the views approach actually introduces more code at call sites, since a view object needs to be created before accessing tokens. More importantly, the NullUtil pattern is transparent and self-explanatory, any developer can immediately see that we are asserting a value to be non-null. With views, the complexity is moved behind an abstraction layer. A new contributor encountering MethodDefView.of(ast) now needs to understand what view classes are, why they exist, and which one to use for a given AST type. That is a steeper learning curve than understanding a single utility method.

I think we should just keep using NullUtil.notNull() and adopt it across the codebase gradually. The proof of concept has already demonstrated that this works well, the code remains readable, and it aligns with the project's emphasis on clarity and maintainability. The pattern NullUtil.notNull(ast.findFirstToken(TokenTypes.IDENT)) clearly communicates intent and satisfies the Checker Framework without adding architectural complexity. We can adopt this pattern file by file, gradually reducing the suppressions list without requiring a large scale refactoring effort.

**Comment by romani** [comment]
We can create PR to experience pain on massive wrapping.

We use pitest and diff reports on massive real sources, so we already protected from crazy coding.

Damaging code to satisfy some tool is not very good.
If we find great balance it will be awesome.

Checker framework is great tool is you have not extensive testing, we are in different state of codebase.

**Comment by Gaoyan1999** [comment]
Understood. It seems the wrapper approach may be too heavy.
I’ll start by using NullUtil.notNull() on a few findFirstToken usages and see how that works in practice.
If it looks good, would it be ok to open the PR under this ticket?

**Comment by romani** [comment]
@Gaoyan1999 , please give us feedback on https://github.com/checkstyle/checkstyle/issues/18420 there is a usage. We appreciate your open honest and direct feedback

**Comment by Gaoyan1999** [comment]
> [@Gaoyan1999](https://github.com/Gaoyan1999) , please give us feedback on [#18420](https://github.com/checkstyle/checkstyle/issues/18420) there is a usage. We appreciate your open honest and direct feedback

commented, LGTM

**Comment by romani** [comment]
@Gaoyan1999 , my bad I referenced wrong resource 
Please review https://github.com/checkstyle/checkstyle/pull/18312

**Comment by Gaoyan1999** [comment]
@romani  Sorry, I actually finished the review yesterday, but I forgot to submit it, so it was pending and only visible to me. I’ve submitted it now—everyone should be able to see it.

**Comment by Gaoyan1999** [comment]
Hi, @vivek-0509  @romani.
When I scan `findFirstToken(Ident)`, I found most of them need NonUtil.notNull. 

<img width="660" height="525" alt="Image" src="https://github.com/user-attachments/assets/50ec392d-a5f7-4621-be3f-51412d9f14bd" />

Is it a good design to add a until function somewhere? We can call this function, when we exactly know the type of ast. Plus, I think this also a good place to `tell` other contributors, which kind of node have IDENT.

The pseudo code below:
```java
   /**
     * Must get the nameAst, if not throw an error.
     */
    public static DetailAST getIdent(DetailAST ast) {
        final Set<Integer> allowedTypes = Set.of(METHOD_DEF, CLASS_DEF, etc...);
        if (!allowedTypes.contains(ast.getType())) {
            throw new AssertionError("Type" + ast + "doesn't have Ident");
        }
        return NullUntils.notNull(ast.findFirstToken(TokenTypes.IDENT));
    }
```

**Comment by vivek-0509** [comment]
@Gaoyan1999, Thanks for the suggestion but my thoughts remain the same as my previous comment. This proposal is essentially NullUtil.notNull() with an extra allowedTypes guard, but that guard only provides a slightly better error message, since NullUtil.notNull() would already throw AssertionError if the wrong node type is passed. Plus, maintaining a hardcoded set of AST types (with conditional cases like parameters) adds complexity without meaningful benefit.

Regarding the documentation aspect, I do agree that it would be valuable to have proper documentation on how we use Checker Framework in our codebase, covering good practices and the utility methods (like NullUtil.notNull()) that developers can use. That would be a better way to guide contributors than encoding this knowledge in code.
